### DISCUSSION TITLE
Cancel "Change Email" Request

### DISCUSSION THREAD

**Discussion by caleb531** [supabase/supabase#7208] [Questions] [answered]
Hi,

I am trying to build Change Email functionality in my React+NextJS app using `supabase.auth.update({ email: ... })` (which sets the `new_email` field on the User object).

```js
await supabase.auth.update({ email: fields.email });
```

However, if the user wants to cancel the request, how can I handle this for them programmatically? I have already tried `supabase.auth.update({ email_change_token: undefined });` to no avail. I wish I could specify `new_email: null` or something, but `new_email` is not a valid property on `UserAttributes`. Is there a supported way of cancelling this pending email-change request in Supabase?

Thanks so much in advance!
Caleb

**Comment by nattstack** [comment]
An idea is using `const { data, error } = await supabase.auth.api.resetPasswordForEmail('user@email.com')` instead on the server and 

When the user clicks the reset link in the email they will be forwarded to: `<SITE_URL>#access_token=x&refresh_token=y&expires_in=z&token_type=bearer&type=recovery` and you can programmatically handle valid email reset with the URL query params on the client.

https://supabase.com/docs/reference/javascript/auth-api-resetpasswordforemail

**Comment by silentworks** [comment]
Does this currently affect the user using their original email to log in? if not I guess you could just ignore the email change request. Another option would be to create a `.rpc` function to set that field to null.

**Comment by caleb531** [comment]
@silentworks Ah, didn't think about an RPC function. That is probably the way to go. Thanks for that suggestion!

I'll give it a shot and let you know what the outcome is.

**Comment by caleb531** [comment, marked as answer]
@silentworks So the solution ended up being twofold: I had to call the RPC function with `supabase.rpc`, then subsequently call the `supabase.auth.update()` function with either an empty object or `{ email_change_token: '' }`. I assume this is because the `update()` call fetches the latest user data from the database every time it's called. If you don't do that, then the front end never knows the database has updated.

Here's the RPC function I wrote:

```sql
-- This SQL function is a Supabase RPC function used for cancelling the process
-- to change the user's email
create or replace function cancel_email_change()
returns json
language plpgsql
security definer
as $$
DECLARE
_uid uuid; -- for checking by 'is not found'
user_id uuid; -- to store the user id from the request
BEGIN

  -- Get user by his current auth.uid
  user_id := auth.uid();

  -- Then clear the relevant fields on the auth.users table
  UPDATE auth.users SET
    email_change = '',
    email_change_token_current = '',
    email_change_token_new = '',
    email_change_confirm_status = 0,
    email_change_sent_at = NULL
  WHERE id = user_id;

  RETURN '{ "success": true }';
END;
$$
```

And again, here's my simple `supabase.rpc` call to the above RPC function with the subsequent call to `supabase.auth.update()`:

```js
const { error } await supabase.rpc('email_change_token');
// Handle errors if I want
await supabase.auth.update({}); // could be anything in this object - not sure it matters
```

**Reply by andrzejkupczyk** [reply]
Also, you might want to consider updating the current email identity (if the user has one), as it will still point to the new email instead of the original one (it's a generated column).

Nevertheless, I must confess that I am not particularly pleased about the prospect of making modifications to these tables...

```sql
    update auth.identities
    set identity_data['email'] = to_jsonb(users.email)
    from (select email
          from auth.users
          where id = auth.uid()) as users
    where user_id = users.id
      and provider = 'email';
```

**Comment by NordlingDev** [comment]
I'm reviving this issue. From a GDPR perspective, what does it mean storing an email address which you don't own and cannot verify, either added by mistake or by malicious reason? This email will be stored in the database and the user has no way of clearing it. And for us developers we have no documented way of clearing it either since we have to go around it by unsafely touching internal tables.
